testsupport/wait/diff.go

```diff
@@ -0,0 +1,93 @@
+package wait
+
+import (
+	"bytes"
+	"strings"
+	"time"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/sergi/go-diff/diffmatchpatch"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func init() {
+	spew.Config.DisablePointerAddresses = true
+	spew.Config.DisableCapacities = true
+}
+
+func Diff(actual, expected interface{}) string {
+	actual = ResetTimeFields(actual)
+	expected = ResetTimeFields(expected)
+	dmp := diffmatchpatch.New()
+	actualdmp, expecteddmp, dmpStrings := dmp.DiffLinesToChars(spew.Sdump(actual), spew.Sdump(expected))
+	diffs := dmp.DiffMain(actualdmp, expecteddmp, false)
+	diffs = dmp.DiffCharsToLines(diffs, dmpStrings)
+	return diffPrettyText(diffs)
+}
+
+// DiffPrettyText converts a []Diff into a colored text report.
+func diffPrettyText(diffs []diffmatchpatch.Diff) string {
+	var buff bytes.Buffer
+
+	_, _ = buff.WriteString("\x1b[31m")
+	_, _ = buff.WriteString("-actual\n")
+	_, _ = buff.WriteString("\x1b[0m")
+	_, _ = buff.WriteString("\x1b[32m")
+	_, _ = buff.WriteString("+expected\n")
+	_, _ = buff.WriteString("\x1b[0m")
+	for _, diff := range diffs {
+		text := diff.Text
+
+		switch diff.Type {
+		case diffmatchpatch.DiffInsert:
+			_, _ = buff.WriteString("\x1b[32m")
+			_, _ = buff.WriteString(strings.Replace(text, " ", "+", 1))
+			_, _ = buff.WriteString("\x1b[0m")
+		case diffmatchpatch.DiffDelete:
+			_, _ = buff.WriteString("\x1b[31m")
+			_, _ = buff.WriteString(strings.Replace(text, " ", "-", 1))
+			_, _ = buff.WriteString("\x1b[0m")
+		case diffmatchpatch.DiffEqual:
+			_, _ = buff.WriteString(text)
+		}
+	}
+
+	return buff.String()
+}
+
+func ResetTimeFields(obj interface{}) interface{} {
+	switch obj := obj.(type) {
+	case []toolchainv1alpha1.Condition:
+		result := make([]toolchainv1alpha1.Condition, len(obj))
+		for i, c := range obj {
+			result[i] = ResetTimeFields(c).(toolchainv1alpha1.Condition)
+		}
+		return result
+	case toolchainv1alpha1.Condition:
+		return toolchainv1alpha1.Condition{
+			Type:               obj.Type,
+			Status:             obj.Status,
+			Reason:             obj.Reason,
+			Message:            obj.Message,
+			LastTransitionTime: metav1.NewTime(time.Time{}),
+		}
+	case []toolchainv1alpha1.UserAccountStatusEmbedded:
+		result := make([]toolchainv1alpha1.UserAccountStatusEmbedded, len(obj))
+		for i, c := range obj {
+			result[i] = ResetTimeFields(c).(toolchainv1alpha1.UserAccountStatusEmbedded)
+		}
+		return result
+	case toolchainv1alpha1.UserAccountStatusEmbedded:
+		return toolchainv1alpha1.UserAccountStatusEmbedded{
+			Cluster:   obj.Cluster,
+			SyncIndex: obj.SyncIndex,
+			UserAccountStatus: toolchainv1alpha1.UserAccountStatus{
+				Conditions: ResetTimeFields(obj.UserAccountStatus.Conditions).([]toolchainv1alpha1.Condition),
+			},
+		}
```

Comment:
only those containing conditions, so we don't have diffs on timestamps, which would be just extra noise

Reply:
ok, but we use conditions probably in all our CRs, right? So is it really too bad to see the diffs of the timestamps? I mean - is it worth it?

just an idea - if we transform the object into a YAML format and then into a generic map of `map[interface{}]interface{}` couldn't we then transitively reset these timestamps in a generic way?

Reply:
I think it's simpler to keep it this way for now, and adapt the code when other resources with conditions are compared. Or we could use the `reflect` package to make things more generic, but at the cost of maintainability.
We could also just ignore this part, but then we'll have extra noise each time there's a diff, and I think we should do something to avoid it.

Reply:
> We could also just ignore this part, but then we'll have extra noise each time there's a diff, and I think we should do something to avoid it.

If there is no generic way of avoiding it, then - personally - I would accept the extra noise so we don't have to have such rewrites of the entire CR objects. I'm just afraid of the complexity because we would need to do it for all CRs and maintain it as soon as we add/remove/modify any parameter.

Reply:
I think from a maintenance perspective, it's easier to live with the extra noise than to have to keep ensuring we update this function for each type that has a timestamp. If it bothers us enough then we should find a more generic solution that doesn't increase the complexity/maintainability.

Reply:
i agree with Matous, it would be an additional step for maintenance of CRS (not sure about the complexity though). 
I couldn't follow how reflect package adds to the cost of maintainability to compare the two options, could you please break it down for me? ( I tried looked at the reflect package, but maybe i'm missing out the obvious)

Reply:
@ranakan19 with the `reflect` package, we could look at the structure of the given object and look for all fields of type `metav1.Time`. But we also need to look at all nested structs, too (can be `[]interface{}`, `toolchainv1alpha1.UserAccountStatus`, etc.) and that's where the code gets more complicated, I believe.

Reply:
ok, so I'll remove this `ResetTimeFields` for now, and we'll see as time goes by if the diffs on the timestamp causes confusion or not.

Reply:
done in 5e252dd